073-decouple-ca-management-from-end-entity-certificate-management.md

```diff
@@ -0,0 +1,363 @@
+
+### Example flows
+
+When a new cluster is first created the flow is:
+
+1. Either the user or the CaReconciler creates the Cluster CA certificate and key and stores them in the <CLUSTER_NAME>-cluster-ca-cert and <CLUSTER_NAME>-cluster-ca Kubernetes Secrets respectively.
+2. The CaReconciler generates the Cluster Operator certificate and stores it in the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret.
+3. The component reconcilers (KafkaReconciler, ZooKeeperReconciler, EntityOperatorReconciler etc) generate their corresponding EE certificates (signed by the newly generated Cluster CA certificate), store them in Kubernetes Secrets and annotate the Kubernetes Secrets and Pods with the cert and key generations, and (for ZooKeeper and Kafka) the certificate hash.
+
+#### Fig 3: New cluster create flow
+
+![New cluster create flow](./images/073-cert-new-existing.png)
+
+> Fig 3: The numbers in the diagram match the numbers in the full description of the steps.
+> The diagram omits the use of annotations.
+
+When the Cluster CA private key is updated (either because the user has requested for it to be regenerated (while Strimzi is managing the CA) or the user has incremented the generation to indicate they've manually updated the Kubernetes Secret):
+1. If Strimzi is managing the Cluster CA, the CaReconciler regenerates the Cluster CAs certificate and key. The old certificate and key are kept in the respective Kubernetes Secret and renamed to ca-YYYY-MM-DDTHH-MM-SSZ.crt.
+2. The CaReconciler skips updating the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret (since the components still only trust the old CA certificate).
+3. The CaReconciler rolls the ZooKeeper and Kafka nodes to trust the new CA certificates.
+4. The ZooKeeper and Kafka reconcilers:
+   1. Generate new certificates signed by the new CA key.
+   2. Update the cluster-ca-cert-generation, cluster-ca-key-generation and server-cert-hash annotations on the StrimziPodSets.
+   3. During their rolling update step, roll the Kubernetes Pods for ZooKeeper and Kafka to use the new certificates (this is triggered by diffing the server-cert-hash annotation).
+5. The EntityOperator, CruiseControl and KafkaExporter reconcilers generate new certificates signed by the new CA key and roll the Kubernetes Pods to use the new certificates.
+6. A new reconciliation starts.
+7. The CaReconciler updates the <CLUSTER_NAME>-cluster-operator-certs Kubernetes Secret with a new certificate signed by the new CA certificate (since the Kafka and ZooKeeper Kubernetes Pods now have the strimzi.io/cluster-ca-cert-generation and strimzi.io/cluster-ca-key-generation matching the current Cluster CA and therefore the Cluster CA is fully trusted and used).
+8. The CaReconciler removes the old CA certificates and keys that have been named ca-YYYY-MM-DDTHH-MM-SSZ.crt.
+9. The component reconcilers roll their Kubernetes Pods, since the Cluster CA passed in has a flag indicating certs have been removed.
```

Comment:
What do you mean here? Which flag? Is this step needed to remove the old CA from their trust stores?

Reply:
Flag is properly not the best word to use, the Ca Java class has a field called `caCertsRemoved` and a method you can call to find out if in the current reconcile loop any CA certificates have been removed https://github.com/strimzi/strimzi-kafka-operator/blob/main/operator-common/src/main/java/io/strimzi/operator/common/model/Ca.java#L772.

So what I am trying to say here is that the reconcilers get told through this internal mechanism that some certs have been removed.

Reply:
I see, thanks for clarifying.